### PR TITLE
Development

### PR DESCRIPTION
**Pull request by rawdaGastan**
## Issues

- https://github.com/rawdaGastan/dmidecode/issues/1

- will remove main.go after testing

### CODE REVIEW

File: pkg/dmi.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
+package pkg
+
+import (
+	"bufio"
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+type Entry = map[string][]string
```
Comment (xmonader) [inline pkg/dmi.go:10]: loved how you are using alias declarations

File: pkg/dmi.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+type DMIMap = map[string]Entry
+
+type DMIDecoder struct {
+	installed  bool
+	cmdOutput  string
+	decodedMap DMIMap
+}
+
+// NewCheckerService creates new instance from the parser
+func NewDMIDecoder() DMIDecoder {
```
Comment (xmonader) [inline pkg/dmi.go:18]: We could have separated the concerns by passing string as a parameter, and we make it the user problem to get a valid dmidecode result to pass to the function, maybe using another function like `GetDMIDecodeOutput()`

File: pkg/dmi.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+type Entry = map[string][]string
+type DMIMap = map[string]Entry
+
+type DMIDecoder struct {
+	installed  bool
```
Comment (xmonader) [inline pkg/dmi.go]: for separation of concertns, the DMIDecoder should only parse data, into an internal map, it shouldn't know about external software installed e.g `sudo, dmidecode`, it should be a concern of another function, or the user provides it

File: pkg/dmi.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	cmdOutput  string
+	decodedMap DMIMap
+}
+
+// NewCheckerService creates new instance from the parser
+func NewDMIDecoder() DMIDecoder {
+	dmidecode := exec.Command("sudo", "dmidecode")
+	res, _ := dmidecode.Output()
```
Comment (xmonader) [inline pkg/dmi.go]: replace `res` with `out`, typically when people read `res` the assume it's the `status code` or the `return code` which is an `int`, also don't ignore the error, if error happens it needs to be handled or propagated to the user e.g in a container and sudo isn't installed

File: pkg/dmi.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+	if len(res) > 0 {
+		return DMIDecoder{
+			installed:  true,
+			cmdOutput:  string(res),
+			decodedMap: DMIMap{},
+		}
+	} else {
+		fmt.Println("dmidecode is not installed in your device")
```
Comment (xmonader) [inline pkg/dmi.go]: this should be an error to be returned, if for logging let's use `zerolog` or standard go logging

File: pkg/dmi.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		return DMIDecoder{}
+	}
+}
+
+// Decodes the string output of dmidecode
+func (dmi *DMIDecoder) Decode() {
```
Comment (xmonader) [inline pkg/dmi.go]: either have parameter representing dmi output passed to it to make it pure or depend on the parameter passed from `NewDMIDecoder` instead

File: pkg/dmi.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+	if !dmi.installed {
```
Comment (xmonader) [inline pkg/dmi.go]: installed check shouldn't be a concern of the dmidecode parser

File: pkg/dmi.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+				key = keyVal[0]
+				val = []string{keyVal[1]}
+
+			} else {
+
+				key = strings.ReplaceAll(keyVal[0], ":", "")
+				val = []string{}
+
+				if key == "System Power Controls" {
```
Comment (xmonader) [inline pkg/dmi.go]: let's make it generic to handle if the line after `Handle line` is starting with spaces, we ignore it and make a new section, I'm not sure if `System Power Controls` will be the only case

**Review by xmonader — CHANGES_REQUESTED** [review]

**Comment by rawdaGastan** [comment]
lgtm :D

**Pull request merged**